server/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java

```diff
@@ -54,17 +55,24 @@ public int totalOperations() {
         return totalOperations;
     }
 
+    int skippedOperations(){
+        return skippedOperations;
+    }
+
     @Override
     Checkpoint getCheckpoint() {
         return checkpoint;
     }
 
     public Translog.Operation next() throws IOException {
-        if (readOperations < totalOperations) {
-            return readOperation();
-        } else {
-            return null;
+        while (readOperations < totalOperations) {
+            final Translog.Operation operation = readOperation();
+            if (operation.seqNo() <= checkpoint.trimmedAboveSeqNo || checkpoint.trimmedAboveSeqNo < 0) {
```

Comment:
can hard check timmerdAboveSeqNo to `UNASSINGED_SEQ_NO`? I don't like the implicit lower than 0. It's hard to grep for usages like that.